File: client/archive/ios/BriskVTU/Features/Beneficiaries/BeneficiaryViewModel.swift
import Foundation
import SwiftUI
import Combine

@MainActor
final class BeneficiaryViewModel: ObservableObject {
    @Published var beneficiaries: [BeneficiaryResponse] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let api = BeneficiaryAPI.shared
    
    func fetchBeneficiaries() async {
        isLoading = true
        errorMessage = nil
        do {
            self.beneficiaries = try await api.listBeneficiaries()
        } catch {
            self.errorMessage = "Failed to load beneficiaries."
            print("Fetch beneficiaries error: \(error)")
        }
        isLoading = false
    }
    
    func addBeneficiary(name: String, serviceType: String, providerServiceId: String, category: String, metadata: [String: String]? = nil) async {
        isLoading = true
        errorMessage = nil
        
        let request = CreateBeneficiaryRequest(
            name: name,
            serviceType: serviceType,
            providerServiceId: providerServiceId,
            category: category,
            metadata: metadata
        )
        
        do {
            _ = try await api.createBeneficiary(request: request)
            await fetchBeneficiaries()
        } catch {
            self.errorMessage = "Failed to save beneficiary."
            print("Create beneficiary error: \(error)")
        }
        isLoading = false
    }
    
    func updateBeneficiary(id: String, name: String? = nil, isFavorite: Bool? = nil) async {
        isLoading = true
        errorMessage = nil
        
        let request = UpdateBeneficiaryRequest(name: name, isFavorite: isFavorite)
        
        do {
            _ = try await api.updateBeneficiary(id: id, request: request)
            await fetchBeneficiaries()
        } catch {
            self.errorMessage = "Failed to update beneficiary."
            print("Update beneficiary error: \(error)")
        }
        isLoading = false
    }
    
    func deleteBeneficiary(id: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            try await api.deleteBeneficiary(id: id)
            self.beneficiaries.removeAll { $0.id == id }
        } catch {
            self.errorMessage = "Failed to delete beneficiary."
            print("Delete beneficiary error: \(error)")
        }
        isLoading = false
    }
}

File: client/archive/android/app/src/main/java/com/solutioncrafts/BriskVTU/feature/beneficiaries/BeneficiaryViewModel.kt
package com.solutioncrafts.BriskVTU.feature.beneficiaries

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solutioncrafts.BriskVTU.network.APIClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class BeneficiaryViewModel : ViewModel() {
    private val api = APIClient.createService(BeneficiaryApi::class.java)

    private val _beneficiaries = MutableStateFlow<List<BeneficiaryResponse>>(emptyList())
    val beneficiaries: StateFlow<List<BeneficiaryResponse>> = _beneficiaries

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    fun fetchBeneficiaries() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                val list = api.listBeneficiaries()
                _beneficiaries.value = list
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Failed to fetch beneficiaries"
                e.printStackTrace()
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun addBeneficiary(request: CreateBeneficiaryRequest) {
        viewModelScope.launch {
            try {
                api.createBeneficiary(request)
                fetchBeneficiaries() // Refresh the list
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Failed to save beneficiary"
                e.printStackTrace()
            }
        }
    }

    fun updateBeneficiary(id: String, name: String? = null, isFavorite: Boolean? = null) {
        viewModelScope.launch {
            try {
                api.updateBeneficiary(id, UpdateBeneficiaryRequest(name, isFavorite))
                fetchBeneficiaries() // Refresh the list
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Failed to update beneficiary"
                e.printStackTrace()
            }
        }
    }

    fun deleteBeneficiary(id: String) {
        viewModelScope.launch {
            try {
                api.deleteBeneficiary(id)
                _beneficiaries.value = _beneficiaries.value.filter { it.id != id }
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Failed to delete beneficiary"
                e.printStackTrace()
            }
        }
    }
}
